src/Components/Server/src/Circuits/CircuitHost.cs

```diff
@@ -136,6 +139,88 @@ internal void SendPendingBatches()
+
+                await Renderer.Dispatcher.InvokeAsync(() =>
+                {
+                    SetCurrentCircuitHost(this);
+                    var document = JsonDocument.Parse(arguments);
+                    if (document.RootElement.ValueKind != JsonValueKind.Array)
+                    {
+                        // Log error and return
+                        return;
+                    }
+                    var length = document.RootElement.GetArrayLength();
+                    if (length != 3)
+                    {
+                        // Log error and return
+                        return;
+                    }
+                    long asyncHandle = 0;
+                    bool succeeded = false;
+                    JSAsyncCallResult result = null;
+                    var i = 0;
+                    foreach (var element in document.RootElement.EnumerateArray())
+                    {
+                        if (i == 0)
+                        {
+                            asyncHandle = element.GetInt64();
+                        }
+                        if (i == 1)
+                        {
+                            succeeded = element.GetBoolean();
+                        }
+
+                        if (i == 2 && element.ValueKind != JsonValueKind.Null)
+                        {
+                            result = (JSAsyncCallResult)typeof(JSAsyncCallResult).GetConstructor(
+                                BindingFlags.NonPublic | BindingFlags.Instance,
+                                binder: null,
+                                new Type[] { typeof(JsonDocument), typeof(JsonElement) },
+                                null).Invoke(new object[] { document, element });
+                        }
+
+                        i++;
+                    }
+
+                    DotNetDispatcher.EndInvoke(asyncHandle, succeeded, result);
+                });
+            }
+            catch (Exception ex)
+            {
+                var args = JsonSerializer.Serialize(
+                    new object[]
+                    {
+                        0,
+                        false,
+                        JSRuntime.SanitizeJSInteropException(ex, typeof(JSRuntimeBase).Assembly.GetName().Name, nameof(DotNetDispatcher.EndInvoke))
+                    },
+                    JsonSerializerOptionsProvider.Options);
+                var _ = JSRuntime.InvokeAsync<object>("DotNet.jsCallDispatcher.endInvokeDotNetFromJS", args);
+            }
+        }
+
+        internal async Task DispatchEvent(RendererRegistryEventDispatcher.BrowserEventDescriptor eventDescriptor, string eventArgsJson)
+        {
+            try
+            {
+                AssertInitialized();
+                await Renderer.Dispatcher.InvokeAsync(() =>
+                {
+                    SetCurrentCircuitHost(this);
+                    return RendererRegistryEventDispatcher.DispatchEvent(eventDescriptor, eventArgsJson);
```

Comment:
Same deal. There's no more need for a magic static in this case.

Reply:
We can dispatch the event directly to the renderer, we would need to have a method to interpret the event args. https://github.com/aspnet/AspNetCore/blob/a784f4575b738ec5b7c6f5cfff2b1fee10de5118/src/Components/Web/src/RendererRegistryEventDispatcher.cs#L24-L25